gam/clustering.py

```diff
@@ -791,3 +510,386 @@ def kmedoids_run(
     def predict(self, X):
         raise NotImplementedError()
 
+    def _update(self, count, mean, m2, new_values):
+        """ Batch updates mu and sigma for bandit PAM using Welford's algorithm
+        Refs:
+            https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+            https://stackoverflow.com/questions/56402955/whats-the-formula-for-welfords-algorithm-for-variance-std-with-batch-updates
+
+        Args:
+            count (int): The number of reference points.
+            mean (int): The current mean
+            m2 (int): The updated mean
+            new_values (np.ndarray): The distance matrix
+
+        Returns:
+            count (int): The number of reference points.
+            mean (int): The old mean.
+            m2 (int): The new mean.
+        """
+        count += len(new_values)
+        # newvalues - oldMean
+        delta = new_values - mean
+        mean += np.sum(delta / count)
+        # newvalues - newMean
+        delta2 = new_values - mean
+        m2 += np.sum(delta * delta2)
+
+        return count, mean, m2
+
+
+    def _finalize(self, count, mean, m2):
+        """Finding variance for each new mean
+
+        Args:
+            count (int): The number of reference points.
+            m2 (int): The updated mean.
+
+        Returns:
+            variance (int): The variance of the medoids
+            TODO: Update this
+        """
+        mean = mean
+        variance = (m2 / count)
+        sample_variance = m2 / (count - 1)
+        if count < 2:
+            return float("nan")
+        else:
+            return mean, variance, sample_variance
+
+   
+    def _bandit_search_singles(self, X, dist_func, d_nearest, td, tmp_arr, j, i):
+        """Inner loop for pam build and bandit build functions.
```

Comment:
Inner loop for both `PAM` and `BanditPAM`? The name suggests this is for `Bandit` only.